otter/convergence/service.py

```diff
@@ -91,6 +91,38 @@ def update_group_state(group, old_state):
 
 
 @do
+def _execute_steps(steps):
+    """
+    Given a set of steps, executes them, logs the result, and returns the worst
+    priority with a list of reasons for that result.
+
+    :return: a tuple of (:class:`StepResult` constant., list of reasons)
+    """
+    if len(steps) > 0:
+        results = yield steps_to_effect(steps)
+
+        severity = [StepResult.FAILURE, StepResult.RETRY, StepResult.SUCCESS]
+        priority = sorted(results,
+                          key=lambda (status, reasons): severity.index(status))
+        worst_status = priority[0][0]
+        results_to_log = zip(
+            steps,
+            [(result, map(structure_reason, reasons))
+             for result, reasons in results])
+
+        reasons = reduce(operator.add,
+                         (x[1] for x in results if x[0] == worst_status))
```

Comment:
To the reviewer:  I wasn't sure why we wanted to pass all possible reasons to `cf_err` so I just went with only returning reasons specifically associated with the worst status.

I know the cloudfeeds logging only logs exceptions it recognizes, but if there happened to be something that sometimes results in a RETRY but is a recognizable exception, we might not want to log that.  I can't think of any right now, though.